cli/internal/oauth/api/api.go

```diff
@@ -0,0 +1,229 @@
+				if *res.Error == "authorization_pending" {
+					continue
+				}
+
+				return res, errors.New(res.ErrorDescription)
+			}
+
+			return res, nil
+		case <-timeout:
+			ticker.Stop()
+			return TokenResponse{}, ErrTimeout
+		}
+	}
+}
+
+// getToken calls the token endpoint of Auth0 and returns the response.
+func (a API) getDeviceToken(ctx context.Context, state State) (TokenResponse, error) {
+	data := url.Values{
+		"client_id":   {a.ClientID},
+		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
+		"device_code": {state.DeviceCode},
+	}
+	oauthTokenURL := a.TenantURL + "/oauth/token"
+
+	resp, err := postForm(ctx, oauthTokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return TokenResponse{}, fmt.Errorf("failed to get tokens: %w", err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	// this endpoint returns a 403 with an `authorization_pending` error until the
+	// user has authenticated, so we don't check the status code here and instead
+	// decode the response and check for the error.
+	var res TokenResponse
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return res, fmt.Errorf("failed to decode response: %w", err)
+	}
+
+	return res, nil
+}
+
+// RevokeToken revokes a refresh token with the tenant so that it can no longer
+// be used to get new tokens.
+func (a API) RevokeToken(ctx context.Context, refreshToken string) error {
+	data := url.Values{
+		"client_id": {a.ClientID},
+		"token":     {refreshToken},
+	}
+
+	revokeURL := a.TenantURL + "/oauth/revoke"
+	resp, err := postForm(ctx, revokeURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return tryDecodeOAuthError(resp)
+	}
+
+	return nil
+}
+
+func postForm(ctx context.Context, reqURL string, data io.Reader) (*http.Response, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, reqURL, data)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	cliVersion := strings.ReplaceAll(version.Version, ".", "_")
+	req.Header.Set("User-Agent", fmt.Sprintf("docker-cli:%s:%s-%s", cliVersion, runtime.GOOS, runtime.GOARCH))
+
+	return http.DefaultClient.Do(req)
+}
+
+func (a API) GetAutoPAT(ctx context.Context, audience string, res TokenResponse) (string, error) {
+	patURL := audience + "/v2/access-tokens/desktop-generate"
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, patURL, nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+res.AccessToken)
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusCreated {
+		return "", fmt.Errorf("unexpected response from Hub: %s", resp.Status)
+	}
+
+	var response patGenerateResponse
```

Comment:
This struct could be inlined, as it's not really used outside of this function
```suggestion
	var response struct {
	    Data struct {
		    Token string `json:"token"`
		}
	}
```

Reply:
Ah, it's actually used in tests!